### ISSUE TITLE
Optional group id & not flag

### ISSUE THREAD

**Issue by finnagin** [RTXteam/RTX#1119]
Combining #1113 and #1112 after the hackathon today.

The google doc from today's hackathon can be found [here.](https://docs.google.com/document/d/15Xo6KufZ6hr42gah-gVS2Uuwo6oeGckF5z20GcR9gLA/edit)

Ideas from hackathon:
* handle not flags inside expand to reduce overall work
* create `option_group_id` which will indicate a group of nodes/edges which are either all in or all out.
* if `option_group_id` is `None` then that edge/node is required
* should always be at least one required path between the source and target nodes.

**Comment by finnagin** [comment]
I created a branch called issue1119 to work on this in.

**Comment by amykglen** [comment]
alright - handling of "not" qedges is implemented in the `issue1119` branch! it appears to be functioning well in testing, but let me know if anyone runs into issues.

one note: while we were originally thinking that both qedges and qnodes should have an `exclude` property, I realized that it may really only make sense for this property to exist on qedges: if a qnode has `exclude=true`, its attached qedge(s) must also be excluded (otherwise we would end up with orphaned edges). and if a qnode isn't attached to any qedges, then I don't think assigning `exclude=true` to it really means anything (seems like a kind of nonsensical query).

so, I suppose there are two options here:
1. make `exclude=true` a property only of qedges (and not qnodes)
2. have code that makes sure any qedge attached to an `exclude=true` qnode also has `exclude=true` set

I favor option 1 for its simpler DSL, but I'm alright with option 2 as well. (`expand`/`resultify` will work as is in either of these paradigms.)

next up is handling `option_groups`... :)

**Comment by dkoslicki** [comment]
I agree that option #1 makes the most sense. If you don’t want a node, then don’t add it in the DSL, basically.

**Comment by finnagin** [comment]
Yeah option 1 makes sense to me too. I'll remove exclude from the add_qnode options and the node swagger model. 

Thanks Amy!

**Comment by dkoslicki** [comment]
@amykglen Since this is now implemented, is there a plan for rolling this out to master + prod?

**Comment by amykglen** [comment]
@dkoslicki - so just the "not" flag portion of this issue is complete. the "option group ID" portion isn't. we could do an intermediary merge to `master` if you want the "not" flag functionality accessible. (I haven't started yet on the "option group ID" implementation, so it would be easy to merge at this point.) otherwise I was figuring we'd wait to merge to `master` until the "option group ID" work is done. (which should be within the next week!)

**Comment by amykglen** [comment]
ok, so I think `option_group_id` functionality is almost fully working now in the `issue1119` branch. here's an example query:
```
add_qnode(id=n00, curie=DOID:3312)
add_qnode(id=n01, type=chemical_substance)
add_qnode(id=n02, type=protein, option_group_id=1, is_set=true)
add_qedge(id=e00, source_id=n00, target_id=n01, type=related_to)
add_qedge(id=e01, source_id=n00, target_id=n02, option_group_id=1, type=predisposes)
add_qedge(id=e02, source_id=n02, target_id=n01, option_group_id=1, type=physically_interacts_with)
expand()
overlay(action=compute_ngd,default_value=inf,virtual_relation_label=N1,source_qnode_id=n00,target_qnode_id=n01)
resultify()
filter_results(action=limit_number_of_results,max_results=10,prune_kg=true)
``` 

these are the pieces remaining that I'm aware of:

1. currently the system only works if optional nodes have `is_set=true` (or have a curie specified) - this can be changed of course, but I'm not totally clear on what we would want results to look like when optional nodes have `is_set=false`... what about if there are multiple option groups, all of which have `is_set=false` for all nodes? are there actual use cases where we'd want to be able to set `is_set=false` on optional nodes?
2. every time a virtual qedge is added, its `option_group_id` now needs to be filled out - that is, if either of its two qnodes is optional (has an `option_group_id`), it should be labeled with that same `option_group_id`. I did this in `compute_ngd` ([here](https://github.com/RTXteam/RTX/commit/2dc53c6e9952f5fad8b964654d02df786209dfcc)), but I think the other `overlay` modules will need to be updated to do this as well.
3. I need to add some more tests/improve existing tests in `test_ARAX_expand.py` and `test_ARAX_resultify.py`

items 1 and 2 might be good topics for a minihackathon!

**Comment by finnagin** [comment]
After hackathon today we discussed that I should add an error/warning when an optional id node is added with is_set=false.

**Comment by amykglen** [comment]
@edeutsch @finnagin @saramsey - following up on yesterday's mini-hackathon: I thought of a way to still handle `not` edges in `expand` (vs. `resultify`), but allow a DSL user to expand edges in whatever order they want: by storing the info for `not`  edges that `expand` finds matches for in a kind of 'hidden' variable on the message, expand can "remember" _all_ previously found `not` edges (from prior `expand()` calls), and thus can check after every edge it expands if there are any paths it needs to blow away.

so, for example, for this query:
```
add_qnode(name=DOID:3312, id=n00)
add_qnode(type=chemical_substance, id=n01)
add_qedge(source_id=n00, target_id=n01, type=indicated_for, id=e00)
add_qedge(source_id=n00, target_id=n01, type=contraindicated_for, exclude=True, id=e01)
expand(kp=ARAX/KG1, edge_id=e00)
expand(kp=ARAX/KG1, edge_id=e01)
resultify()
```
one could alternatively do the two expand steps in this order:
```
expand(kp=ARAX/KG1, edge_id=e01)
expand(kp=ARAX/KG1, edge_id=e00)
```
and the same result set is produced in the end.

so, in summary, we still get the performance benefit of handling "not" edges in `expand` vs. `resultify`, but also don't introduce any restrictions for DSL users.

this change is in the `issue1119` branch.

**Comment by edeutsch** [comment]
sounds great!

**Comment by dkoslicki** [comment]
@finnagin can you add a few examples in test_ARAX workflows? In particular, include DTD+not known to treat and not contraindicated.

**Comment by amykglen** [comment]
hey @finnagin - I noticed that it looks like `add_qnode()` is forcing `is_set=true` for qnodes with an `option_group_id` even if the qnode also has a `curie`. I think it should only force `is_set=true` if the optional qnode doesn't have a curie specified?

**Comment by finnagin** [comment]
Oops! I'll fix that now. Thanks for catching that @amykglen !

**Comment by finnagin** [comment]
Ok @amykglen , it will now only force `is_set=true` when both the `curie` and `name` fields are not set.

**Comment by amykglen** [comment]
awesome, thanks @finnagin!

and I just adjusted each `overlay` module so that they all fill out `QEdge.option_group_id` when adding a virtual qedge to the QG. so I think everything should be in place for this issue! I'm just planning to play around with a few more test queries to make sure everything seems good...

**Comment by amykglen** [comment]
ok, this has been merged into `master`. if people want to play around with the functionality, here are a few example queries:

one-hop query with a "not" (aka, "exclude") edge:
```
add_qnode(id=n00, name=DOID:3312)
add_qnode(id=n01, type=chemical_substance)
add_qedge(id=e00, source_id=n00, target_id=n01, type=indicated_for)
add_qedge(id=eXX, source_id=n00, target_id=n01, type=contraindicated_for, exclude=true)
expand(kp=ARAX/KG1)
resultify()
```

one-hop query with a single optional edge:
```
add_qnode(id=n00, curie=DOID:3312)
add_qnode(id=n01, type=chemical_substance)
add_qedge(id=e00, source_id=n00, target_id=n01, type=positively_regulates)
add_qedge(id=group1_e1, source_id=n00, target_id=n01, type=correlated_with, option_group_id=1)
expand(kp=ARAX/KG2)
resultify()
```

query where the required portion is one-hop and there's a two-hop optional group:
```
add_qnode(id=n00, curie=DOID:3312)
add_qnode(id=n01, curie=CHEBI:48607)
add_qnode(id=group1_n1, type=protein, option_group_id=1, is_set=true)
add_qedge(id=e00, source_id=n00, target_id=n01, type=related_to)
add_qedge(id=group1_e1, source_id=n00, target_id=group1_n1, option_group_id=1, type=predisposes)
add_qedge(id=group1_e2, source_id=group1_n1, target_id=n01, option_group_id=1, type=physically_interacts_with)
expand()
resultify()
```

query where the required portion is one-hop and there's a two-hop optional group that also has a "not" edge:
```
add_qnode(id=n0, curie=DOID:3312)
add_qnode(id=n1, curie=CHEBI:48607)
add_qnode(id=group1_n1, type=protein, option_group_id=1, is_set=true)
add_qnode(id=group1_n2x, curie=UMLS:C0023692, option_group_id=1)
add_qedge(id=e0, source_id=n0, target_id=n1, type=related_to)
add_qedge(id=group1_e1, source_id=n0, target_id=group1_n1, option_group_id=1, type=predisposes)
add_qedge(id=group1_e2, source_id=group1_n1, target_id=n1, option_group_id=1, type=physically_interacts_with)
add_qedge(id=group1_ex, source_id=group1_n2x, target_id=group1_n1, option_group_id=1, exclude=true)
expand()
resultify()
```

and there are more example queries here and here.

note from AHM today: Eric pointed out that the query graph interpreter will likely need to be updated so that it knows what to do when it sees a query graph with "not" edges or optional paths. but that isn't super urgent.
